### PR TITLE
Add compat data for CanvasRenderingContext2D

### PR DESCRIPTION
**Pull request by maboa**
Phew! That was tough going. I hope I've nested the data correctly, when it comes to multiple rows in the compat data - eg https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/addHitRegion

### CODE REVIEW

File: api/CanvasRenderingContext2D.json [hunk 1]
```diff
@@ -0,0 +1,3857 @@
...
+      },
+      "filter": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/filter",
+          "support": {
+            "webview_android": {
+              "version_added": true
+            },
+            "chrome": {
+              "version_added": "52"
```
Comment (jpmedley) [inline api/CanvasRenderingContext2D.json]: I suspect that it should be 52 for both webview and chrome android. I'm verifying this with the [project owners](https://www.chromestatus.com/features/6701649301078016) and will get back to you.

Rule of thumb: if Webview and Chrome Android version numbers are missing from a Chrome Status entry, it's more likely that the project owner neglected to fill in the boxes than that it is that the feature is not supported. Email the owner if you have questions. That's why email addresses are on Chrome Status.

Comment (jpmedley) [reply api/CanvasRenderingContext2D.json]: Or contact me through my GitHub account.

File: api/CanvasRenderingContext2D.json [hunk 2]
```diff
@@ -0,0 +1,3857 @@
...
+            "deprecated": false
+          }
+        }
+      },
+      "imageSmoothingEnabled": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/imageSmoothingEnabled",
+          "support": {
+            "webview_android": {
+              "version_added": null
```
Comment (jpmedley) [inline api/CanvasRenderingContext2D.json]: This should also be "30".

File: api/CanvasRenderingContext2D.json [hunk 3]
```diff
@@ -0,0 +1,3857 @@
...
+            "deprecated": false
+          }
+        }
+      },
+      "imageSmoothingQuality": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/imageSmoothingQuality",
+          "support": {
+            "webview_android": {
+              "version_added": null
```
Comment (jpmedley) [inline api/CanvasRenderingContext2D.json]: This should also be 54.

**Comment by jpmedley** [comment]
PLEASE DON'T MERGE YET. I'm reviewing the Chrome stuff.

File: api/CanvasRenderingContext2D.json (hunk 1 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: OK @jpmedley - thanks. Waiting to hear back from you on this one before making the appropriate changes and pushing.

**Comment by maboa** [comment]
Note - I just spotted what I think is a misplaced curly bracket (or lack thereof) near the top of the file. Will attempt to remedy in the next push.

**Comment by Elchi3** [comment]
I haven't started in depth review, but identifiers should be without braces even if they are methods. 
`"addHitRegion()": {` -> `"addHitRegion": {`

So that when accessing the data, it will be 
`compat.api.CanvasRenderingContext2D.addHitRegion;` instead of
`compat.api.CanvasRenderingContext2D['addHitRegion()'];`.

**Comment by maboa** [comment]
OK that's fine Elchi3 - I thought I saw it elsewhere and followed suit, but happy to fix. 👍

File: api/CanvasRenderingContext2D.json (hunk 1 above)
Comment (jpmedley) [reply api/CanvasRenderingContext2D.json]: I just pinged them again. If I don't get an answer before @Elchi3 finishes his review you are free to merge and I'll submit a separate PR later.

Comment (jpmedley) [reply api/CanvasRenderingContext2D.json]: I just got an answer. This is in both Chrome Android and Android Webview.

File: api/CanvasRenderingContext2D.json (hunk 2 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: @jpmedley - Both Chrome and Webview Android, or just Webview?

File: api/CanvasRenderingContext2D.json (hunk 3 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: @jpmedley - Both Chrome and Webview Android, or just Webview?

File: api/CanvasRenderingContext2D.json [hunk 4]
```diff
@@ -0,0 +1,3857 @@
...
+        },
+        "Returns a DOMMatrix": {
+          "__compat": {
+            "support": {
+              "webview_android": {
+                "version_added": null
+              },
+              "chrome": {
+                "version_added": false,
+                "notes": "See <a href='https://crbug.com/637940'>https://crbug.com/637940</a>."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: Updated.

File: api/CanvasRenderingContext2D.json (hunk 1 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: Updated.

File: api/CanvasRenderingContext2D.json (hunk 2 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: Presumed just Webview for now.

File: api/CanvasRenderingContext2D.json (hunk 3 above)
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: Presumed just Webview for now.

File: api/CanvasRenderingContext2D.json [hunk 5]
```diff
@@ -0,0 +1,3859 @@
...
+      "addHitRegion": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/addHitRegion",
+          "support": {
+            "webview_android": {
+              "version_added": false
+            },
+            "chrome": {
+              "version_added": true,
+              "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Please see `currentTransform`, this needs a flags object.

File: api/CanvasRenderingContext2D.json [hunk 6]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "edge": {
+              "version_added": null
+            },
+            "edge_mobile": {
+              "version_added": null
+            },
+            "firefox": {
+              "version_added": "30",
+              "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: This needs to be a `flags` object.

File: api/CanvasRenderingContext2D.json [hunk 7]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "firefox_android": {
+              "version_added": "30"
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: ditto here

File: api/CanvasRenderingContext2D.json [hunk 8]
```diff
@@ -0,0 +1,3859 @@
...
+        },
+        "id": {
+          "__compat": {
+            "support": {
+              "webview_android": {
+                "version_added": false
+              },
+              "chrome": {
+                "version_added": true,
+                "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 9]
```diff
@@ -0,0 +1,3859 @@
...
+              },
+              "edge": {
+                "version_added": null
+              },
+              "edge_mobile": {
+                "version_added": null
+              },
+              "firefox": {
+                "version_added": "30",
+                "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 10]
```diff
@@ -0,0 +1,3859 @@
...
+        },
+        "control": {
+          "__compat": {
+            "support": {
+              "webview_android": {
+                "version_added": false
+              },
+              "chrome": {
+                "version_added": true,
+                "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 11]
```diff
@@ -0,0 +1,3859 @@
...
+              },
+              "edge": {
+                "version_added": null
+              },
+              "edge_mobile": {
+                "version_added": null
+              },
+              "firefox": {
+                "version_added": "30",
+                "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 12]
```diff
@@ -0,0 +1,3859 @@
...
+        },
+        "path": {
+          "__compat": {
+            "support": {
+              "webview_android": {
+                "version_added": false
+              },
+              "chrome": {
+                "version_added": true,
+                "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 13]
```diff
@@ -0,0 +1,3859 @@
...
+              },
+              "edge": {
+                "version_added": null
+              },
+              "edge_mobile": {
+                "version_added": null
+              },
+              "firefox": {
+                "version_added": "39",
+                "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 14]
```diff
@@ -0,0 +1,3859 @@
...
+        },
+        "fillRule": {
+          "__compat": {
+            "support": {
+              "webview_android": {
+                "version_added": false
+              },
+              "chrome": {
+                "version_added": true,
+                "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 15]
```diff
@@ -0,0 +1,3859 @@
...
+              "safari": {
+                "version_added": false
+              },
+              "safari_ios": {
+                "version_added": false
+              }
+            }
+          }
+        },
+        "other hit region options": {
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: No identifier with spaces here. Please use underscores, and add `"description": "Other hit region options"`.

File: api/CanvasRenderingContext2D.json [hunk 16]
```diff
@@ -0,0 +1,3859 @@
...
+      "clearHitRegions": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/clearHitRegions",
+          "support": {
+            "webview_android": {
+              "version_added": false
+            },
+            "chrome": {
+              "version_added": true,
+              "notes": "This feature is behind a feature flag. Set the flag <code>ExperimentalCanvasFeatures</code> to <code>true</code> to enable it."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 17]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "edge": {
+              "version_added": null
+            },
+            "edge_mobile": {
+              "version_added": null
+            },
+            "firefox": {
+              "version_added": "38",
+              "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 18]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "firefox_android": {
+              "version_added": "38",
+              "notes": "This feature is behind a feature preference setting. In about:config, set <code>canvas.hitregions.enabled</code> to <code>true</code>."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

File: api/CanvasRenderingContext2D.json [hunk 19]
```diff
@@ -0,0 +1,3859 @@
...
+              "version_added": true
+            }
+          },
+          "status": {
+            "experimental": false,
+            "standard_track": true,
+            "deprecated": false
+          }
+        },
+        "Path parameter": {
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "\<code>Path\</code> parameter".

File: api/CanvasRenderingContext2D.json [hunk 20]
```diff
@@ -0,0 +1,3859 @@
...
+            "edge": {
+              "version_added": null
+            },
+            "edge_mobile": {
+              "version_added": null
+            },
+            "firefox": [
+              {
+                "version_added": "29",
+                "notes": "In Firefox 28 this method was implemented as <code>drawSystemFocusRing</code>, but has been renamed to the standard <code>drawFocusIfNeeded</coded> in Firefox 29."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: This is quite complex :)
The first range should be { version_added: 32 }
The second range should be { version_added: 29, flag }
The third range should be { version_added: 28, flag, alternative_name: drawSystemFocusRing }

File: api/CanvasRenderingContext2D.json [hunk 21]
```diff
@@ -0,0 +1,3859 @@
...
+              },
+              {
+                "version_added": "32",
+                "notes": "Prior to Firefox 32 this method was deactivated by default behind the flag <code>canvas.focusring.enabled</code>."
+              }
+            ],
+            "firefox_android": [
+              {
+                "version_added": "29",
+                "notes": "In Firefox 28 this method was implemented as <code>drawSystemFocusRing</code>, but has been renamed to the standard <code>drawFocusIfNeeded</coded> in Firefox 29."
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: ditto here

File: api/CanvasRenderingContext2D.json [hunk 22]
```diff
@@ -0,0 +1,3859 @@
...
+              "version_added": true
+            }
+          },
+          "status": {
+            "experimental": false,
+            "standard_track": true,
+            "deprecated": false
+          }
+        },
+        "Path parameter": {
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "\<code>Path\</code> parameter".

File: api/CanvasRenderingContext2D.json [hunk 23]
```diff
@@ -0,0 +1,3859 @@
...
+              "version_added": true
+            }
+          },
+          "status": {
+            "experimental": false,
+            "standard_track": true,
+            "deprecated": false
+          }
+        },
+        "ImageBitmap as source image": {
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could be "ImageBitmap_source_image", and description "\<code>ImageBitmap\</code> as source image".

File: api/CanvasRenderingContext2D.json [hunk 24]
```diff
@@ -0,0 +1,3859 @@
...
+                "version_added": null
+              },
+              "opera_android": {
+                "version_added": null
+              },
+              "safari": {
+                "version_added": null
+              },
+              "safari_ios": {
+                "vers
```
Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Please fix identifier, see above.

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Please fix identifier, see above.

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "<code>Path</code> parameter".

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: getImage -> getImageData

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "\<code>Path\</code> parameter".

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "\<code>Path\</code> parameter".

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: `flags` object needed

Comment (Elchi3) [inline api/CanvasRenderingContext2D.json]: Same here, no white space. Identifier could just be "path", and description "\<code>Path\</code> parameter".

File: api/CanvasRenderingContext2D.json [hunk 25]
```diff
@@ -0,0 +1,3859 @@
...
+      "currentTransform": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/currentTransform",
+          "support": {
+            "webview_android": {
+              "version_added": false
+            },
+            "chrome": {
+              "version_added": true,
+              "notes": "This feature is deactivated by default. Enable the <code>ExperimentalCanvasFeatures</code> flag to test it."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json (hunk 3 above)
Comment (jpmedley) [reply api/CanvasRenderingContext2D.json]: Both.

File: api/CanvasRenderingContext2D.json [hunk 26]
```diff
@@ -0,0 +1,3859 @@
...
+              "version_added": false
+            }
+          },
+          "status": {
+            "experimental": true,
+            "standard_track": true,
+            "deprecated": false
+          }
+        },
+        "Returns a DOMMatrix": {
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 27]
```diff
@@ -0,0 +1,3859 @@
...
+                "version_added": null
+              },
+              "edge_mobile": {
+                "version_added": true
+              },
+              "firefox": {
+                "version_added": null
+              },
+              "firefox_android": {
+                "version_added": "31"
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 28]
```diff
@@ -0,0 +1,3859 @@
...
+      "direction": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/direction",
+          "support": {
+            "webview_android": {
+              "version_added": true
+            },
+            "chrome": {
+              "version_added": true,
+              "notes": "This feature is deactivated by default. Enable the <code>ExperimentalCanvasFeatures</code> flag to test it."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 29]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "edge": {
+              "version_added": null
+            },
+            "edge_mobile": {
+              "version_added": null
+            },
+            "firefox": {
+              "version_added": "49",
+              "notes": "The <code>filter</code> property was deactivated by default in versions 35-48. Set the flag <code>canvas.filters.enabled</code> to <code>true</code> in these versions. Starting with Firefox 41, the initial value is none, see (bug(1163124))."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 30]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "firefox_android": {
+              "version_added": "49",
+              "notes": "The <code>filter</code> property was deactivated by default in versions 35-48. Set the flag <code>canvas.filters.enabled</code> to <code>true</code> in these versions. Starting with Firefox 41, the initial value is none, see (bug(1163124))."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: :+1:

File: api/CanvasRenderingContext2D.json [hunk 31]
```diff
@@ -0,0 +1,3859 @@
...
+            },
+            "safari": {
+              "version_added": true
+            },
+            "safari_ios": {
+              "version_added": true
+            }
+          },
+          "status": {
+            "experimental": true,
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 32]
```diff
@@ -0,0 +1,3859 @@
...
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

File: api/CanvasRenderingContext2D.json [hunk 33]
```diff
@@ -0,0 +1,3859 @@
...
+      "imageSmoothingEnabled": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CanvasRenderingContext2D/imageSmoothingEnabled",
+          "support": {
+            "webview_android": {
+              "version_added": "30"
+            },
+            "chrome": {
+              "version_added": "30",
+              "notes": "Supported with a <code>webkit</code> prefix until version 29."
```
[the remark this answers is not in the document]
Comment (maboa) [reply api/CanvasRenderingContext2D.json]: 👍

**Comment by maboa** [comment]
Hopefully I've addressed all the issues! Thanks for the comprehensive review.

**Comment by Elchi3** [comment]
Thanks for addressing the review comments so thoroughly. We are getting very close. 

From the GitHub UI it looks like two things are missed:
- .canvas is not yet experimental: false
- .clearHitRegions still needs the flags object

Also, as said above, you forgot [fillText](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fillText) and [scale](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/scale), but I would also be happy to handle these in a follow-up. This is super large already :)

Thanks again for your hard work on this.

**Comment by maboa** [comment]
No worries. Hopefully the bits I missed have now been addressed :)

**Pull request merged**